src/string_id.h

```diff
@@ -69,69 +108,92 @@ template<typename T>
 class string_id
 {
     public:
+        template<class U>
+        friend class string_id;
         using This = string_id<T>;
 
-        /**
-         * Forwarding constructor, forwards any parameter to the std::string
-         * constructor to create the id string. This allows plain C-strings,
-         * and std::strings to be used.
-         */
-        // Beautiful C++11: enable_if makes sure that S is always something that can be used to constructor
-        // a std::string, otherwise a "no matching function to call..." error is generated.
+        template<unsigned N>
+        explicit constexpr string_id( const char ( &str )[N] ) : _hash( Hash::CompileTime::fnv1_64( str ) ),
+            _char_ptr( str ), _is_char_ptr( true ) {}
+
+        template<typename U>
+        string_id( const string_id<U> &other ) : _hash( other.hash() ), _char_ptr( other._char_ptr ),
```

Comment:
Does this mean you can convert string ids of one object type to that of another object type? I don't think we should allow that.

Reply:
Added it cause I saw a `string_id<X>(string_id<Y>.str())`, its a way to do the same without calculating the hash again. If it shouldn't be done, I can find those and remove them.

Reply:
There are some places in the code that do this, but I don't think they're common enough to be performance-relevant, and we should probably be moving away from this anyway, in favour of e.g. explicitly linking two potentially-different values in JSON.